internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -257,14 +257,13 @@ func (r *Reconciler) computeDesiredMachineSet(ctx context.Context, deployment *c
 		name = existingMS.Name
 		uid = existingMS.UID
 
-		// Keep foregroundDeletion finalizer if the existingMS has it.
-		// Note: This case is a little different from the create case. In the update case we preserve
-		// the finalizer on the MachineSet if it already exists. Because of SSA we should not build
-		// the finalizer information from the MachineDeployment when updating a MachineSet because that could lead
-		// to dropping the finalizer from the MachineSet if it is dropped from the MachineDeployment.
-		// We should not drop the finalizer on the MachineSet if the finalizer is dropped from the MachineDeployment.
-		if sets.New[string](existingMS.Finalizers...).Has(metav1.FinalizerDeleteDependents) {
-			finalizers = []string{metav1.FinalizerDeleteDependents}
+		// Preserve all existing finalizers (including foregroundDeletion finalizer).
+		// Ensure MachineSet has foregroundDeletion finalizer if the MachineDeployment has it.
+		// We should not drop the foregroundDeletion finalizer on the MachineSet if the finalizer is dropped from the MachineDeployment.
+		finalizers = existingMS.Finalizers
+		if sets.New[string](deployment.Finalizers...).Has(metav1.FinalizerDeleteDependents) &&
```

Comment:
Not clear to me if we should propagate the foregroundDeletion finalizer at all (in MD & MS)